File: Chapter9ZeusStudy1_3/pluginapi/src/main/java/com/mars/pluginapi/PluginManager.java
package com.mars.pluginapi;

import android.app.Application;
import android.content.Context;
import android.content.res.AssetManager;
import android.content.res.Resources;

import com.mars.components.util.file.FileCopyUtils;
import com.mars.components.util.reflect.FieldUtils;

import java.io.File;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

import kotlin.jvm.Volatile;

/**
 * Created by geyan on 2021/1/19
 */
public class PluginManager {

    public final static List<PluginItem> plugins = new ArrayList<>();

    // ContextImpl中的LoadedApk对象中的mPackageInfo
    private static Object mPackageInfo;

    // 原始的application中的BaseContext，不能是其他的，否则会内存泄漏
    @Volatile
    private static Context mBaseContext;

    @Volatile
    public static Resources mCurResources;


    public static void init(Application application) {
        mBaseContext = application.getBaseContext();
        mPackageInfo = FieldUtils.readField(mBaseContext, "mPackageInfo");

        mCurResources = mBaseContext.getResources();

        try {
            AssetManager assetManager = application.getAssets();
            String[] assetsList = assetManager.list("");
            // 遍历宿主的assets文件件，拷贝插件apk
            for (String path : assetsList) {
                if (path.endsWith(".apk")) {
                    String apkName = path;
                    String dexName = apkName.replace(".apk", ".dex");

                    FileCopyUtils.extractAssets(mBaseContext, apkName);

                    mergeDexs(apkName, dexName);
                    PluginItem item = generatePluginItem(apkName);
                    plugins.add(item);
                }
            }
            reloadInstallPluginResources(plugins);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void mergeDexs(String apkName, String dexName) {

        File dexFile = mBaseContext.getFileStreamPath(apkName);
        File optDexFile = mBaseContext.getFileStreamPath(dexName);

        try {
            BaseDexClassLoaderHookHelper.mergeClassLoader(mBaseContext.getClassLoader(), dexFile, optDexFile);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }


    private static void reloadInstallPluginResources(List<PluginItem> plugins) {

        try {
            AssetManager assetManager = AssetManager.class.newInstance();
            Method addAssetPath = AssetManager.class.getMethod("addAssetPath", String.class);

            addAssetPath.invoke(assetManager, mBaseContext.getPackageResourcePath());

            for (PluginItem item : plugins) {
                addAssetPath.invoke(assetManager, item.pluginPath);
            }

            Resources newResources = new Resources(assetManager,
                    mBaseContext.getResources().getDisplayMetrics(),
                    mBaseContext.getResources().getConfiguration());

            FieldUtils.writeField(mBaseContext, "mResources", newResources);
            //这是最主要的需要替换的，如果不支持插件运行时更新，只留这一个就可以了
            FieldUtils.writeField(mPackageInfo, "mResources", newResources);

            //清除一下之前的resource的数据，释放一些内存
            //因为这个resource有可能还被系统持有着，内存都没被释放
            //clearResoucesDrawableCache(mNowResources);

            mCurResources = newResources;
            //需要清理mtheme对象，否则通过inflate方式加载资源会报错
            //如果是activity动态加载插件，则需要把activity的mTheme对象也设置为null
            FieldUtils.writeField(mBaseContext, "mTheme", null);
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }

    private static PluginItem generatePluginItem(String apkName) {
        File file = mBaseContext.getFileStreamPath(apkName);
        PluginItem item = new PluginItem();
        item.pluginPath = file.getAbsolutePath();
        item.packageInfo = DLUtils.getPackageInfo(mBaseContext, item.pluginPath);
        return item;
    }

}

File: ZeusStudy1_1/pluginapi/src/main/java/com/mars/pluginapi/PluginManager.kt
package com.mars.pluginapi

import android.app.Application
import android.content.Context
import android.content.res.AssetManager
import android.content.res.Resources

/**
 * Created by geyan on 2020/11/6
 */
object PluginManager {

    private val plugins = arrayListOf<PluginItem>()

    // 原始的application的BaseContext，不能是其他的，否则会内存泄露
    @Volatile
    private lateinit var mBaseContext: Context

    // 当前正在使用的Resources
    @Volatile
    private lateinit var mCurResources: Resources

    // ContextImpl中的LoadedApk对象的mPackageInfo
    private lateinit var mPackageInfo: Any

    fun init(application: Application) {
        mBaseContext = application.baseContext
        mCurResources = mBaseContext.resources
        mPackageInfo = RefInvoke.getFieldObject(mBaseContext, "mPackageInfo")

        try {
            val assetManager = application.assets
            val paths = assetManager.list("")
            val pluginPaths = arrayListOf<String>()

            for (path: String in paths!!) {
                if (path.endsWith(".apk")) {
                    val apkName = path
                    val dexName = apkName.replace(".apk", ".dex")
                    Utils.extractAssets(mBaseContext, apkName)
                    mergeDexs(apkName, dexName)
                    with(generatePluginItem(apkName)) {
                        plugins.add(this)
                        pluginPaths.add(this.pluginPath)
                    }
                }
            }
            reloadInstalledPluginResources(pluginPaths)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun generatePluginItem(apkName: String): PluginItem {
        val file = mBaseContext.getFileStreamPath(apkName)
        return PluginItem(Utils.getPackageInfo(mBaseContext, file.absolutePath), file.absolutePath)
    }

    private fun mergeDexs(apkName: String, dexName: String) {
        val dexFile = mBaseContext.getFileStreamPath(apkName)
        val optDexFile = mBaseContext.getFileStreamPath(dexName)

        BaseDexClassLoaderHookHelper.patchClassLoader(mBaseContext.classLoader, dexFile, optDexFile)
    }

    private fun reloadInstalledPluginResources(pluginPaths: ArrayList<String>) {
        try {
            val assetManager = AssetManager::class.java.newInstance()
            val addAssetPath =
                AssetManager::class.java.getMethod("addAssetPath", String::class.java)
            addAssetPath.invoke(assetManager, mBaseContext.packageResourcePath)
            for (pluginPath: String in pluginPaths) {
                addAssetPath.invoke(assetManager, pluginPath)
            }
            val newResources = Resources(
                assetManager,
                mBaseContext.resources.displayMetrics,
                mBaseContext.resources.configuration
            )
            RefInvoke.setFieldObject(mBaseContext, "mResources", newResources)
            // 这是最主要的需要替换的，如果不支持插件运行时更新，只留这一个即可
            RefInvoke.setFieldObject(mPackageInfo, "mResources", newResources)
            mCurResources = newResources
            // 需要清理mTheme对象，否则通过inflate方式加载资源会报错
            // 如果是Activity动态加载插件，则需要把activity的mTheme对象也设置为null
            RefInvoke.setFieldObject(mBaseContext, "mTheme", null)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun getPlugins(): ArrayList<PluginItem> {
        return plugins
    }

    fun getResources(): Resources {
        return mCurResources
    }
}